zerver/lib/users.py

```diff
@@ -348,6 +349,34 @@ def access_user_by_email(
     return access_user_common(target, user_profile, allow_deactivated, allow_bots, for_admin)
 
 
+def access_user_by_delivery_email(
+    user_profile: UserProfile,
+    email: str,
+    *,
+    allow_deactivated: bool = False,
+    allow_bots: bool = False,
+    for_admin: bool,
+) -> UserProfile:
+    """
+    Checks if user_profile can access the target user by .delivery_email address. Therefore, only
+    returns the fetched UserProfile if the requester is allowed to see the target's email address
+    (on top of passing the other access_user_common checks).
+
+    The purpose of this is to be used at API endpoints that allow selecting the target user order_by
+    delivery_email, while preventing the endopoint from leaking information about user emails.
+
+    TODO: **Warning**: This function is vulnerable to timing attacks, as it doesn't do the
+    user fetch + can_access_delivery_email check in constant time.
```

Comment:
This is the main caveat of this PR to highlight imo. Perhaps this means that we should at least protect this new endpoint behind `@require_realm_owner`?